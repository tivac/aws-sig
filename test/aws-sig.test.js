"use strict";

const config = require("./lib/config.js");

const { signedHeaders : sign } = require("../src/aws-sig.js");

describe("aws-sig core API", () => {
    it("should sort SignedHeaders by name", () => {
        expect(
            sign({
                url     : "https://aws.amazon.com",
                headers : {
                    "X-Amz-Date" : "20150830T123600Z",

                    z : 1,
                    a : 1,
                    m : 1,
                },
            }, config())
        ).toMatchSnapshot();
    });

    it("should throw on a missing request param", () => {
        expect(() =>
            sign()
        ).toThrowErrorMatchingSnapshot();
    });

    it("should throw on missing config param", () => {
        expect(() =>
            sign({})
        ).toThrowErrorMatchingSnapshot();
    });

    it("should throw on a missing config param values", () => {
        expect(() =>
            sign({ url : "https://aws.amazon.com" }, {})
        ).toThrowErrorMatchingSnapshot();
    });
    
    it("should be precise about the missing config param values", () => {
        const conf = config();

        delete conf.accessKeyId;
        delete conf.region;
        
        expect(() =>
            sign({ url : "https://aws.amazon.com" }, conf)
        ).toThrowErrorMatchingSnapshot();
    });

    it("should clean up paths with relative directory specifiers", () => {
        expect(
            sign({
                url : "https://aws.amazon.com/foo/./bar/../baz",
            }, config()).test.canonical
        ).toMatchSnapshot();
    });
    
    it("should clean up paths with multiple slashes", () => {
        expect(
            sign({
                url : "https://aws.amazon.com/foo/////bar",
            }, config()).test.canonical
        ).toMatchSnapshot();
    });

    it.each([
        "/foo bar",
        "/foo%20bar",
        "/%20",
        "/%2a",
        "/%41",
        "/Ã¼",
        "/arn%3Aaws%3Aservice%3Aus-west-2%3A%3Aident%2Fid1%2Fid2",
        "/foo*",
    ])("should double-encode each path segment (%s)", (path) => {
        expect(
            sign({
                url : `https://aws.amazon.com${path}`,
            }, config()).test.canonical
        ).toMatchSnapshot();
    });

    it("should leave S3 paths alone", () => {
        const conf = config();

        conf.service = "s3";
        
        expect(
            sign({
                url : `https://aws.amazon.com/s3//allows//for//weird.paths`,
            }, conf).test.canonical
        ).toMatchSnapshot();
    });
});
