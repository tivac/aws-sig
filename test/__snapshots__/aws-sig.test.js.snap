// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws-sig core API should be precise about the missing config param values 1`] = `"Missing required config fields: accessKeyId, region"`;

exports[`aws-sig core API should clean up paths with multiple slashes 1`] = `
"GET
/foo/bar




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should clean up paths with relative directory specifiers 1`] = `
"GET
/foo/baz




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/%2a) 1`] = `
"GET
/%252a




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/%20) 1`] = `
"GET
/%2520




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/%41) 1`] = `
"GET
/%2541




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/arn%3Aaws%3Aservice%3Aus-west-2%3A%3Aident%2Fid1%2Fid2) 1`] = `
"GET
/arn%253Aaws%253Aservice%253Aus-west-2%253A%253Aident%252Fid1%252Fid2




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/foo bar) 1`] = `
"GET
/foo%2520bar




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/foo%20bar) 1`] = `
"GET
/foo%2520bar




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/foo*) 1`] = `
"GET
/foo%2A




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should double-encode each path segment (/Ã¼) 1`] = `
"GET
/%25C3%25BC




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should leave S3 paths alone 1`] = `
"GET
/s3//allows//for//weird.paths




e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
`;

exports[`aws-sig core API should sort SignedHeaders by name 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Authorization": "AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request, SignedHeaders=a;m;x-amz-date;z, Signature=d5b729b777af569c5e22fbf1d1a27fad8d490b6fc3c7a5e1a54542bfa68c0998",
    "X-Amz-Date": "20150830T123600Z",
    "a": 1,
    "m": 1,
    "z": 1,
  },
  "method": "GET",
  "test": Object {
    "auth": "AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request, SignedHeaders=a;m;x-amz-date;z, Signature=d5b729b777af569c5e22fbf1d1a27fad8d490b6fc3c7a5e1a54542bfa68c0998",
    "canonical": "GET
/

a:1
m:1
x-amz-date:20150830T123600Z
z:1

a;m;x-amz-date;z
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "sts": "AWS4-HMAC-SHA256
20150830T123600Z
20150830/us-east-1/service/aws4_request
3d94ead6a8f2253deb74d7183e308aaa8c311c7f8248b19b69667f65639de28c",
  },
  "url": "https://aws.amazon.com/",
}
`;

exports[`aws-sig core API should throw on a missing config param values 1`] = `"Missing required config fields: accessKeyId, region, secretAccessKey, service"`;

exports[`aws-sig core API should throw on a missing request param 1`] = `"Missing request object"`;

exports[`aws-sig core API should throw on missing config param 1`] = `"Missing config object"`;
